### PR TITLE
host: Perform explicit check for running containers

### PR DESCRIPTION
**Pull request by lmars**
I keep looking at CI build failures and finding #907 to be the cause, so this is an attempt at a workaround.

### CODE REVIEW

**Comment by titanous** [comment]
Hmm, any ideas why the running event isn't arriving?

**Comment by lmars** [comment]
> Hmm, any ideas why the running event isn't arriving?

No idea, I have looked quite a few times at it and got nowhere, so decided to just add this hack.

I haven't got as far as sysdig as it is hard to reproduce, but happens in approximately 1 in 5 failed CI jobs.

File: host/libvirt_lxc_backend.go [hunk 1]
```diff
@@ -778,41 +778,64 @@ func (c *libvirtContainer) watch(ready chan<- error, buffer host.LogBuffer) erro
...
+				c.Client.Resume()
+				c.l.state.SetStatusFailed(c.job.ID, errors.New("container failed to start"))
+				return nil
+			}
+		case <-runningCheck:
+			if isRunning {
+				continue
+			}
+			g.Log(grohl.Data{"at": "running_check"})
+			if state, err := c.Client.GetState(); err == nil && state == containerinit.StateRunning {
```
Comment (titanous) [inline host/libvirt_lxc_backend.go]: This API call is cheap, why not do it once a second from the beginning?

Comment (lmars) [inline host/libvirt_lxc_backend.go]: Yep makes sense, updated in 841e809

**Comment by lmars** [comment]
@titanous hmm it seems `c.Client.GetState` just blocks indefinitely, and I can't see it called anywhere else so maybe it has never worked? I am not overly familiar with the RPC code, any thoughts?

**Comment by lmars** [comment]
@titanous found the cause of the hang, I was creating multiple state stream channels and not listening on them which blocks the rpcplus loop [here](https://github.com/flynn/flynn/blob/master/pkg/rpcplus/client.go#L181-L185).

Fixed in d9ae843, please review.

File: host/libvirt_lxc_backend.go [hunk 2]
```diff
@@ -778,41 +778,66 @@ func (c *libvirtContainer) watch(ready chan<- error, buffer host.LogBuffer) erro
...
-			g.Log(grohl.Data{"at": "change", "status": "error", "err": err})
-			c.Client.Resume()
-			c.l.state.SetStatusFailed(c.job.ID, err)
-			return err
+	// containerinit sometimes fails to send the StateRunning state change
+	// when the container is running, so we do an explicit check once a
+	// second to avoid the job being stuck in the "starting" state (see
+	// https://github.com/flynn/flynn/issues/907)
+	isRunning := false
+	runningCheck := time.Tick(time.Second)
```
Comment (titanous) [inline host/libvirt_lxc_backend.go]: This should be a `time.Ticker` and get stopped when finished with it.

File: host/libvirt_lxc_backend.go [hunk 3]
```diff
@@ -778,41 +778,66 @@ func (c *libvirtContainer) watch(ready chan<- error, buffer host.LogBuffer) erro
...
+				return nil
+			case containerinit.StateFailed:
+				g.Log(grohl.Data{"at": "failed"})
+				c.Client.Resume()
+				c.l.state.SetStatusFailed(c.job.ID, errors.New("container failed to start"))
+				return nil
+			}
+		case <-runningCheck:
+			if isRunning {
+				runningCheck = nil
```
Comment (titanous) [inline host/libvirt_lxc_backend.go]: Could just nil it out in `handleRunning`, then `isRunning` would not be required.

**Comment by titanous** [comment]
LGTM

**Pull request merged**